### ISSUE TITLE
Declare Manually DI to work

### ISSUE THREAD

**Issue by jonathanperis** [Cysharp/MessagePipe#35]
Hello!

I'm trying to use the MessagePie on my project but i'm only able to sucefully use it if i declare by myself the DI linking the AsyncRequestHandlers to their respective Handlers.

I'm asking this because i don't have saw on the docs that declare the DI by myself is a required step to make MessagePie works.

So i've did: 

```csharp
            services
                .AddSingleton<IAsyncRequestHandler<CreateSistemaCommand, CreateSistemaResponse>, SistemaHandler>()
                .AddSingleton<IAsyncRequestHandler<ListSistemaQuery, ListSistemaResponse>, SistemaHandler>()
                .AddSingleton<IAsyncRequestHandler<GetSistemaQuery, GetSistemaResponse>, SistemaHandler>()
                .AddSingleton<IAsyncRequestHandler<RemoveSistemaCommand, RemoveSistemaResponse>, SistemaHandler>()
                .AddSingleton<IAsyncRequestHandler<UpdateSistemaCommand, UpdateSistemaResponse>, SistemaHandler>();
```

If i don't declare by myself, the code throws an exception when i try to use any of the non declared DIs:

```csharp
System.InvalidOperationException: Unable to resolve service for type 'MessagePipe.IAsyncRequestHandlerCore`2[Cpnucleo.Infra.CrossCutting.Util.Queries.Sistema.ListSistema.ListSistemaQuery,Cpnucleo.Infra.CrossCutting.Util.Queries.Sistema.ListSistema.ListSistemaResponse]' while attempting to activate 'MessagePipe.AsyncRequestHandler`2[Cpnucleo.Infra.CrossCutting.Util.Queries.Sistema.ListSistema.ListSistemaQuery,Cpnucleo.Infra.CrossCutting.Util.Queries.Sistema.ListSistema.ListSistemaResponse]'.
```

The question is: The MessagePipe should be handle by ifself the DIs (like Mediatr does) or it's necessary declare on the client side?

Here's my solution: https://github.com/jonathanperis/cpnucleo/tree/feat-message-pipe-hellyeah

The projects are the **Cpnucleo.GRPC** that contains the mediator like classes to invoke their handlers contained on the **Cpnucleo.Application** project

Classes used by their handlers are shared on the **Cpnucleo.Infra.CrossCutting.Util** project

**Comment by jonathanperis** [comment]
So after dig some situations i got what's went wrong:

If you set more than one RequestHandler injector on the constructor (on my situation this will be extremely normal since i group the gRPC calls by the CRUD class) the error will happen:

> public class SistemaGrpcService : ServiceBase<ISistemaGrpcService>, ISistemaGrpcService
    {
        private readonly IAsyncRequestHandler<CreateSistemaCommand, CreateSistemaResponse> _createSistemaCommand;
        private readonly IAsyncRequestHandler<ListSistemaQuery, ListSistemaResponse> _listSistemaQuery;
        private readonly IAsyncRequestHandler<GetSistemaQuery, GetSistemaResponse> _getSistemaQuery;
        private readonly IAsyncRequestHandler<RemoveSistemaCommand, RemoveSistemaResponse> _removeSistemaCommand;
        private readonly IAsyncRequestHandler<UpdateSistemaCommand, UpdateSistemaResponse> _updateSistemaCommand;

        public SistemaGrpcService(IAsyncRequestHandler<CreateSistemaCommand, CreateSistemaResponse> createSistemaCommand, 
                                  IAsyncRequestHandler<ListSistemaQuery, ListSistemaResponse> listSistemaQuery, 
                                  IAsyncRequestHandler<GetSistemaQuery, GetSistemaResponse> getSistemaQuery, 
                                  IAsyncRequestHandler<RemoveSistemaCommand, RemoveSistemaResponse> removeSistemaCommand, 
                                  IAsyncRequestHandler<UpdateSistemaCommand, UpdateSistemaResponse> updateSistemaCommand)
        {
            _createSistemaCommand = createSistemaCommand;
            _listSistemaQuery = listSistemaQuery;
            _getSistemaQuery = getSistemaQuery;
            _removeSistemaCommand = removeSistemaCommand;
            _updateSistemaCommand = updateSistemaCommand;
        }

        public async UnaryResult<CreateSistemaResponse> AddAsync(CreateSistemaCommand command)
        {
            return await _createSistemaCommand.InvokeAsync(command);
        }

        public async UnaryResult<ListSistemaResponse> AllAsync(ListSistemaQuery query)
        {
            return await _listSistemaQuery.InvokeAsync(query);
        }

        public async UnaryResult<GetSistemaResponse> GetAsync(GetSistemaQuery query)
        {
            return await _getSistemaQuery.InvokeAsync(query);
        }

        public async UnaryResult<RemoveSistemaResponse> RemoveAsync(RemoveSistemaCommand command)
        {
            return await _removeSistemaCommand.InvokeAsync(command);
        }

        public async UnaryResult<UpdateSistemaResponse> UpdateAsync(UpdateSistemaCommand command)
        {
            return await _updateSistemaCommand.InvokeAsync(command);
        }
    }

for my scenario, can you sugest me a way to solve this? with the MediatR i only have to inject one instance of it but the requests/responses are already been defined on their respective classes (e.g: IRequest<GetSistemaResponse> interface that you need to apply on the GetSistemaQuery class if i use MediatR)

**Comment by neuecc** [comment]
Currently, AutoRegistration can only register one interface per concrete type.
In other words

```csharp
internal class SistemaHandler :
    IAsyncRequestHandler<CreateSistemaCommand, CreateSistemaResponse>,
    IAsyncRequestHandler<GetSistemaQuery, GetSistemaResponse>,
    IAsyncRequestHandler<ListSistemaQuery, ListSistemaResponse>,
    IAsyncRequestHandler<RemoveSistemaCommand, RemoveSistemaResponse>,
    IAsyncRequestHandler<UpdateSistemaCommand, UpdateSistemaResponse>
``

This will register only

```csharp
IAsyncRequestHandler<CreateSistemaCommand, CreateSistemaResponse>, SistemaHandler 
```

This is a bug, I will be fixed, thanks!

--

Your manually registration code is invalid, you need to use `AddAsyncRequestHandler`.
However, this also allows only one type to be registered,
so it was not possible to register a type with multiple handlers like your `SistemaHandler`.

This will also be fixed, due to lack of consideration.

--

I consider a class like `IMediator` to be the same thing as ServiceLocator.
In other words, I think it's not good code.
We should specify the type in the constructor.

In any case, the above fixes should resolve your problem.
Please wait a moment.

**Comment by neuecc** [comment]
I've released `1.4.1`, it includes these fix.
Please try it.

**Comment by jonathanperis** [comment]
I've tried and it's fixed.

Thanks!
